File: test/language/templates/cpp98/InstantiateTypeAsSqlDatabaseFieldTest.cpp
#include <cstdio>

#include "gtest/gtest.h"

#include "templates/instantiate_type_as_sql_database_field/InstantiateTypeAsSqlDatabaseFieldDb.h"

namespace templates
{
namespace instantiate_type_as_sql_database_field
{

class InstantiateTypeAsSqlDatabaseFieldTest : public ::testing::Test
{
public:
    InstantiateTypeAsSqlDatabaseFieldTest()
    {
        std::remove(DB_FILE_NAME.c_str());
    }

    static const std::string DB_FILE_NAME;

protected:
    template <typename T>
    void assertEqualRows(const std::vector<T>& rows1, const std::vector<T>& rows2)
    {
        ASSERT_EQ(rows1.size(), rows2.size());
        for (size_t i = 0; i < rows1.size(); ++i)
        {
            assertEqualRow(rows1.at(i), rows2.at(i));
        }
    }

    template <typename T>
    void assertEqualRow(const T& row1, const T& row2)
    {
        ASSERT_TRUE(row1.getId() == row2.getId());
        ASSERT_TRUE(row1.getData() == row2.getData());
    }
};

const std::string InstantiateTypeAsSqlDatabaseFieldTest::DB_FILE_NAME = "instantiate_type_as_sql_database_field_test.sqlite";

TEST_F(InstantiateTypeAsSqlDatabaseFieldTest, readWrite)
{
    InstantiateTypeAsSqlDatabaseFieldDb instantiateTypeAsSqlDatabaseFieldDb(DB_FILE_NAME);
    instantiateTypeAsSqlDatabaseFieldDb.createSchema();

    StringTable& stringTable = instantiateTypeAsSqlDatabaseFieldDb.getStringTable();
    std::vector<StringTableRow> stringTableRows;
    StringTableRow stringRow1;
    stringRow1.setId(0);
    stringRow1.setData("test");
    stringTableRows.push_back(stringRow1);
    stringTable.write(stringTableRows);

    StringTable& otherStringTable = instantiateTypeAsSqlDatabaseFieldDb.getOtherStringTable();
    std::vector<StringTableRow> otherStringTableRows;
    StringTableRow otherStringRow1;
    otherStringRow1.setId(0);
    otherStringRow1.setData("other test");
    otherStringTableRows.push_back(otherStringRow1);
    otherStringTable.write(otherStringTableRows);

    InstantiateTypeAsSqlDatabaseFieldDb readInstantiateTypeAsSqlDatabaseFieldDb(DB_FILE_NAME);
    std::vector<StringTableRow> readStringTableRows;
    readInstantiateTypeAsSqlDatabaseFieldDb.getStringTable().read(readStringTableRows);
    std::vector<StringTableRow> readOtherStringTableRows;
    readInstantiateTypeAsSqlDatabaseFieldDb.getOtherStringTable().read(readOtherStringTableRows);

    assertEqualRows(stringTableRows, readStringTableRows);
    assertEqualRows(otherStringTableRows, readOtherStringTableRows);
}

} // namespace instantiate_type_as_sql_database_field
} // namespace templates

File: test/language/templates/java/templates/InstantiateTypeAsSqlDatabaseFieldTest.java
package templates;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.IOException;
import java.sql.SQLException;
import java.util.List;
import java.util.ArrayList;

import test_utils.FileUtil;
import test_utils.JdbcUtil;

import zserio.runtime.BitSizeOfCalculator;
import zserio.runtime.ZserioError;
import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.BitStreamWriter;
import zserio.runtime.io.FileBitStreamReader;
import zserio.runtime.io.FileBitStreamWriter;

import templates.instantiate_type_as_sql_database_field.InstantiateTypeAsSqlDatabaseFieldDb;
import templates.instantiate_type_as_sql_database_field.StringTable;
import templates.instantiate_type_as_sql_database_field.StringTableRow;

public class InstantiateTypeAsSqlDatabaseFieldTest
{

    @BeforeAll
    public static void init()
    {
        JdbcUtil.registerJdbc();
    }

    @BeforeEach
    public void setUp()
    {
        FileUtil.deleteFileIfExists(dbFile);
    }

    @Test
    public void readWrite() throws IOException, SQLException
    {
        final InstantiateTypeAsSqlDatabaseFieldDb instantiateTypeAsSqlDatabaseFieldDb =
                new InstantiateTypeAsSqlDatabaseFieldDb(DB_FILE_NAME);
        instantiateTypeAsSqlDatabaseFieldDb.createSchema();

        final StringTable stringTable = instantiateTypeAsSqlDatabaseFieldDb.getStringTable();
        final List<StringTableRow> stringTableRows = new ArrayList<StringTableRow>();
        final StringTableRow stringRow1 = new StringTableRow();
        stringRow1.setId(0);
        stringRow1.setData("test");
        stringTableRows.add(stringRow1);
        stringTable.write(stringTableRows);

        final StringTable otherStringTable = instantiateTypeAsSqlDatabaseFieldDb.getOtherStringTable();
        final List<StringTableRow> otherStringTableRows = new ArrayList<StringTableRow>();
        final StringTableRow otherStringRow1 = new StringTableRow();
        otherStringRow1.setId(0);
        otherStringRow1.setData("other test");
        otherStringTableRows.add(otherStringRow1);
        otherStringTable.write(otherStringTableRows);

        instantiateTypeAsSqlDatabaseFieldDb.close();

        final InstantiateTypeAsSqlDatabaseFieldDb readInstantiateTypeAsSqlDatabaseFieldDb =
                new InstantiateTypeAsSqlDatabaseFieldDb(DB_FILE_NAME);
        final List<StringTableRow> readStringTableRows =
                readInstantiateTypeAsSqlDatabaseFieldDb.getStringTable().read();
        final List<StringTableRow> readOtherStringTableRows =
                readInstantiateTypeAsSqlDatabaseFieldDb.getOtherStringTable().read();

        assertEqualRows(stringTableRows, readStringTableRows);
        assertEqualRows(otherStringTableRows, readOtherStringTableRows);

        readInstantiateTypeAsSqlDatabaseFieldDb.close();
    }

    static void assertEqualRows(List<StringTableRow> rows1, List<StringTableRow> rows2)
    {
        assertEquals(rows1.size(), rows2.size());
        for (int i = 0; i < rows1.size(); ++i)
        {
            assertEqualRow(rows1.get(i), rows2.get(i));
        }
    }

    static void assertEqualRow(StringTableRow row1, StringTableRow row2)
    {
        assertEquals(row1.getId(), row2.getId());
        assertEquals(row1.getData(), row2.getData());
    }

    private static final String DB_FILE_NAME = "instantiate_type_as_sql_database_field.sqlite";
    private final File dbFile = new File(DB_FILE_NAME);
}
